x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java

```diff
@@ -95,29 +116,44 @@ protected void doAuthenticate(UsernamePasswordToken token, ActionListener<Authen
         } else if (ClientReservedRealm.isReserved(token.principal(), config.settings()) == false) {
             listener.onResponse(AuthenticationResult.notHandled());
         } else {
-            getUserInfo(token.principal(), ActionListener.wrap((userInfo) -> {
-                AuthenticationResult result;
+            getUserInfo(token.principal(), (userInfo) -> {
                 if (userInfo != null) {
-                    try {
-                        if (userInfo.hasEmptyPassword()) {
-                            result = AuthenticationResult.terminate("failed to authenticate user [" + token.principal() + "]", null);
-                        } else if (userInfo.verifyPassword(token.credentials())) {
+                    if (userInfo.hasEmptyPassword()) {
+                        listener.onResponse(AuthenticationResult.terminate("failed to authenticate user [" + token.principal() + "]"));
+                    } else {
+                        ActionListener<AuthenticationResult> hashCleanupListener = ActionListener.runBefore(listener, () -> {
+                            if (userInfo != null && userInfo != bootstrapUserInfo && userInfo != autoconfigUserInfo) {
```

Comment:
we check for `userInfo != null` in line 120 so we can remove this from here and make it sliiiiiightly simpler